modules/sdk-lib-mpc/src/tss/ecdsa/noSmallFactorsProof.ts

```diff
@@ -0,0 +1,177 @@
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {nonce, e} - challenge value $e$ and associated $nonce$ that makes $e$ uniformly random from $(-order, order)$.
+ */
+function generateEforProve(N: bigint, w: bigint): { nonce: Buffer; e: bigint } {
+  let nonce: Buffer, e: bigint, digest: Buffer;
+  do {
+    nonce = randomBytes(33);
+    digest = hash(N, w, nonce);
+    e = bigIntFromBufferBE(digest.subarray(1));
+  } while (e >= ORDER);
+  if (digest[0] & 1) {
+    return { nonce, e: -e };
+  }
+  return { nonce, e };
+}
+
+/**
+ * Generate pseudo-random challenge value $e$ for $(N, w)$ and associated $nonce$.
+ * @param N - the prime number to verify is a product of two large primes.
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @param nonce - a random nonce.
+ * @returns {bigint} - challenge value $e$.
+ */
+function generateEforVerify(N: bigint, w: bigint, nonce: Buffer): bigint {
+  const digest = hash(N, w, nonce);
+  const e = bigIntFromBufferBE(digest.subarray(1));
+  if (digest[0] & 1) {
+    return -e;
+  }
+  return e;
+}
+
+/**
+ * Calculate the closest integer square root of $n$.
+ * @param n - the number to calculate the square root of.
+ * @returns {bigint} - $n$'s closest integer square root.
+ */
+function isqrt(n: bigint): bigint {
+  if (n < BigInt(0)) {
+    throw new Error();
+  }
+  if (n < BigInt(2)) {
+    return n;
+  }
+  function newtonIteration(n: bigint, x0: bigint) {
+    const x1 = (n / x0 + x0) >> BigInt(1);
+    if (x0 === x1 || x0 === x1 - BigInt(1)) {
+      return x0;
+    }
+    return newtonIteration(n, x1);
+  }
+  return newtonIteration(n, BigInt(1));
+}
+
+/**
+ * Prove that $n0$ has no small factors, where $n0$ is the product of two large primes.
+ * @param p - a large prime.
+ * @param q - a large prime.
+ * @param w - a random number with the same bitLength as $p * q$, that satisfies the Jacobi of w is -1 wrt $n0$.
+ * @param nHat - a safe bi-prime, such as that returned from rangeProof.generateNTilde.
+ * @param s - security parameters for $nHat$ such as the $h1$ value returned from rangeProof.generateNTilde.
+ * @param t - security parameters for $nHat$ such as the $h2$ value returned from rangeProof.generateNTilde.
+ * @returns proof that the product of $p * q$ has no small factors.
+ */
+export function prove(
+  p: bigint,
+  q: bigint,
+  w: bigint,
+  nHat: bigint,
+  s: bigint,
+  t: bigint
+): DeserializedNoSmallFactorsProof {
+  const n0 = p * q;
+  const { nonce, e } = generateEforProve(n0, w);
+  const sqrtN0 = isqrt(n0);
+  const alpha = randBetween(sqrtN0 << (ELL + EPSILON), -sqrtN0 << (ELL + EPSILON));
+  const beta = randBetween(sqrtN0 << (ELL + EPSILON), -sqrtN0 << (ELL + EPSILON));
+  const rho = randBetween((nHat * n0) << ELL, -(nHat * n0) << ELL);
+  // Commit to p.
+  const mu = randBetween(BigInt(1) << ELL, BigInt(-1) << ELL);
```

Comment:
I think mu should be 
`randBetween(BigInt(nHat) << ELL, BigInt(-nHat) << ELL) `according to https://eprint.iacr.org/2020/492.pdf B.4